rust/lance/src/io/commit.rs

```diff
@@ -771,28 +740,38 @@ pub(crate) async fn commit_transaction(
 
     // Note: object_store has been configured with WriteParams, but dataset.object_store()
     // has not necessarily. So for anything involving writing, use `object_store`.
-    let transaction_file = write_transaction_file(object_store, &dataset.base, transaction).await?;
     let read_version = transaction.read_version;
     let mut target_version = read_version + 1;
-    let mut dataset = dataset.clone();
-    if matches!(transaction.operation, Operation::Overwrite { .. })
+    let original_dataset = dataset.clone();
+
+    let mut dataset = if matches!(transaction.operation, Operation::Overwrite { .. })
         && commit_config.num_retries == 0
     {
-        dataset.checkout_version(transaction.read_version).await?;
+        dataset.checkout_version(transaction.read_version).await?
     } else {
         // We need to checkout the latest version, because any fixes we apply
         // (like computing the new row ids) needs to be done based on the most
         // recent manifest.
+        let mut dataset = dataset.clone();
         dataset.checkout_latest().await?;
-    }
+        dataset
+    };
+
+    let mut transaction = transaction.clone();
+
     let num_attempts = std::cmp::max(commit_config.num_retries, 1);
+    // TODO: use SlotBackoff here instead and size unit off of attempt time.
```

Comment:
This is likely going to be more important with the rebases in here now. Look at the retries in `merge_insert.rs` for how we use `SlotBackoff`.